gaia-header.js

```diff
@@ -34,124 +54,519 @@ module.exports = Component.register('gaia-header', {
+   * after the last <h1>.
+   *
+   * @return {Array}
+   * @private
+   */
+  getButtonsAfterTitle: function() {
+    var children = this.children;
+    var els = [];
+
+    for (var i = children.length - 1; i >= 0; i--) {
+      var el = children[i];
+      if (el.tagName === 'H1') { break; }
+      if (!isButton(el)) { continue; }
+      els.push(el);
+    }
+
+    return els;
+  },
+
+  /**
+   * Get the sum of the width of
+   * the given buttons.
+   *
+   * This function is optimized to avoid reading
+   * `element.clientWidth` when possible.
+   *
+   * If a button is `display: none` it will
+   * have a `.clientWidth` of 0, therefore won't
+   * contribute anything to the overall sum.
+   *
+   * @param  {Array} buttons
+   * @return {Number}
+   * @private
+   */
+  sumButtonWidths: function(buttons) {
+    var defaultWidth = 50;
+    var sum = buttons.reduce((prev, button) => {
+      var isStandardButton = button === this.els.actionButton;
+      var width = isStandardButton ? defaultWidth : button.clientWidth;
+      return prev + width;
+    }, 0);
+
+    debug('sum button widths', buttons, sum);
+    return sum;
+  },
+
+  /**
+   * Known attribute property
+   * descriptors.
+   *
+   * These setters get called when matching
+   * attributes change on the element.
+   *
+   * @type {Object}
+   */
+  attrs: {
+    action: {
+      get: function() { return this._action; },
+      set: function(value) {
+        var action = KNOWN_ACTIONS[value];
+        if (action === this._action) { return; }
+        this.setAttr('action', action);
+        this._action = action;
+      },
+    },
+
+    titleStart: {
+      get: function() {
+        debug('get title-start');
+        if ('_titleStart' in this) { return this._titleStart; }
+        var buttons = this.getButtonsBeforeTitle();
+        var value = this.sumButtonWidths(buttons);
+        debug('get title-start', buttons, value);
+        return value;
+      },
+
+      set: function(value) {
+        debug('set title-start', value);
+        value = parseInt(value, 10);
+        if (value === this._titleStart || isNaN(value)) { return; }
+        this.setAttr('title-start', value);
+        this._titleStart = value;
+        debug('set');
+      }
+    },
+
+    titleEnd: {
+      get: function() {
+        debug('get title-end');
+        if ('_titleEnd' in this) { return this._titleEnd; }
+        var buttons = this.getButtonsAfterTitle();
+        return this.sumButtonWidths(buttons);
+      },
+
+      set: function(value) {
+        debug('set title-end', value);
+        value = parseInt(value, 10);
+        if (value === this._titleEnd || isNaN(value)) { return; }
+        this.setAttr('title-end', value);
+        this._titleEnd = value;
+      }
+    },
+
+    noFontFit: {
+      get: function() { return this._noFontFit || false; },
+      set: function(value) {
+        debug('set no-font-fit', value);
+        value = !!(value || value === '');
```

Comment:
wondering if it's more readable to do `value = value !== null && value !== undefined` (or `value != null` which is the same). And technically more right because you could assign `0` (even if unlikely) :)